File: plot_read_stats.py
home_dir = '/share/ScratchGeneral/jamtor/'
project_dir = home_dir + 'projects/ewing_ctDNA/'
fq_dir = project_dir + 'raw_files/'
result_dir = project_dir + 'results/'
stats_dir = result_dir + 'stats/'
bam_dir = result_dir + 'picard/bams/'

import pysam
import sys
import re
import os

def minmax(val_list):
    min_val = min(val_list)
    max_val = max(val_list)
    return (min_val, max_val)

orig_fq = fq_dir + "409_004_D9YWF_GTAGAGGA-CTCTCTAT_L001_R1.fastq"








#inbam_name = sys.argv[1]
#stats_name = sys.argv[2]
#discordant_name = sys.argv[3]
#split_name = sys.argv[4]
inbam_name = bam_dir + '409_004_D9YWF_GTAGAGGA-CTCTCTAT_L001.initial_mapped_and_UMI.bam'
stats_name = stats_dir + '409_004_D9YWF_GTAGAGGA-CTCTCTAT_L001.uncollapsed.bam.bc'
discordant_name = bam_dir + '409_004_D9YWF_GTAGAGGA-CTCTCTAT_L001.uncollapsed.discordant.bam'
split_name = bam_dir + '409_004_D9YWF_GTAGAGGA-CTCTCTAT_L001.uncollapsed.split.bam'
out_prefix = stats_dir + re.sub('initial_mapped_and_UMI.bam$', '', os.path.basename(inbam_name))

print('Input bam = ', inbam_name)
print('Stats file = ', stats_name)
print('Discordant bam = ', discordant_name)
print('Split bam = ', split_name)
print('Output prefix = ', out_prefix)

# define filename using pysam:
inbam = pysam.Samfile(inbam_name, 'rb')

# define counters for reads processed and written:
n = 0
w = 0

print('Fetching lengths of sequences...')

lengths=[]
for read in inbam.fetch(until_eof=True):
    n += 1
    # add read seq length to lengths list:
    lengths.append(len(read.seq))

# plot length distribution:




	assert umi1 is not None
	# remove last 11 characters:
	umi1_fix = re.sub('...........$', '', umi1)
	# replace original umi:
	read.set_tag('ZB', umi1_fix, value_type='Z')
	# fetch RX tag:
	umi2 = read.get_tag('RX')
	assert umi2 is not None
	# remove last 11 characters:
	umi2_fix = re.sub('[A-Z]-', '', 
		re.sub('...........$', '', umi1)
	)
	# replace original umi:
	read.set_tag('RX', umi2_fix, value_type='Z')

	# if length = 12, write to file:
	if len(umi1_fix) == 12 & len(umi2_fix) == 12:
		if '-' not in umi2_fix:
			w += 1
			outbam.write(read)
		else:
			print('Hyphen still in RX UMI sequence: ', umi2_fix, ' read=', read)
			break
	else:
		print(
			'UMI sequence wrong length: ZB=', umi1_fix, ' RX=', umi2_fix, ' read=', read
		)

print('All reads processed, checked and written')

inbam.close()
outbam.close()

File: plot_read_stats.R

sample_name <- "409_004_D9YWF_GTAGAGGA-CTCTCTAT_L001"

home_dir <- "/share/ScratchGeneral/jamtor/"
project_dir <- paste0(home_dir, "projects/ewing_ctDNA/")
raw_dir <- paste0(project_dir, "raw_files/")
fq_dir <- paste0(result_dir, "picard/fastqs/")
int_dir <- paste0(result_dir, "picard/int_bams/")
bam_dir <- paste0(result_dir, "picard/bams/")
result_dir <- paste0(project_dir, "results/")
stats_dir <- paste0(result_dir, "stats/", sample_name, "/")
temp_dir <- paste0(result_dir, "picard/temp/")

system(paste0("mkdir -p ", stats_dir))
system(paste0("mkdir -p ", temp_dir))


##################################################################################################################################
### 0. Load libraries and functions ###
##################################################################################################################################

library(ShortRead)
library(rtracklayer)
library(GenomicRanges)

mround <- function(x,base){
  base*round(x/base)
}


##################################################################################################################################
### 1. Load data ###
##################################################################################################################################

# load fastqs:
fqs <- list(
  orig_fq = readFastq(paste0(raw_dir, sample_name, "_R1.fastq.gz")),
  filtered_orig_fq = readFastq(paste0(fq_dir, sample_name, ".withUMI.fastq")),
  consensus_fq = readFastq(paste0(fq_dir, sample_name, ".consensus.fastq"))
)


mapped_bams <- list(
  initial_mapped_bam = import(paste0(bam_dir, sample_name, ".uncollapsed.bam")),
  initial_mapped_discordant_bam = import(paste0(bam_dir, sample_name, ".uncollapsed.discordant.bam")),
  initial_mapped_split_bam = import(paste0(bam_dir, sample_name, ".uncollapsed.split.bam")),
  consensus_remapped_bam = import(paste0(bam_dir, sample_name, ".consensus.unfiltered.bam")),
  consensus_remapped_discordant_bam = import(paste0(bam_dir, sample_name, ".consensus.unfiltered.discordant.bam")),
  consensus_remapped_split_bam = import(paste0(bam_dir, sample_name, ".consensus.unfiltered.split.bam")),
  filtered_consensus_bam = import(paste0(bam_dir, sample_name, ".consensus.filtered.bam")),
  filtered_consensus_discordant_bam = import(paste0(bam_dir, sample_name, ".consensus.filtered.discordant.bam")),
  filtered_consensus_split_bam = import(paste0(bam_dir, sample_name, ".consensus.filtered.split.bam"))
)


##################################################################################################################################
### 2. Determine sequence lengths and total read no from original fq1 file and collapsed fq file ###
##################################################################################################################################

fq_lengths <- lapply(fqs, function(x) {
  return(width(sread(x)))
})


##################################################################################################################################
### 3. Determine sequence lengths and total raw and read nos from bams ###
##################################################################################################################################

# determine unmapped bam names:
unmapped_names <- c(".unmapped.bam", ".unmapped.filtered.bam", ".consensus.unmapped.bam")
for (i in 1:length(unmapped_names)) {

  print(i)

  # remove header:
  system(
    paste0(
    	"samtools view ", int_dir, sample_name, unmapped_names[i], " > ", temp_dir, sample_name, "_no_header", 
      unmapped_names[i]
    )
  )
  # load bam:
  df <- read.table(
    paste0(temp_dir, sample_name, "_no_header", unmapped_names[i]),
    sep = "\t",
    fill = TRUE,
    header = FALSE
  )
  	
  # create length list:
  if (i==1) {
    unmapped_bam_lengths <- list(nchar(df$V10))
  } else {
    unmapped_bam_lengths[[i]] <- nchar(df$V10)
  }

}
names(unmapped_bam_lengths) <- gsub(
  "^_", "", 
  gsub("\\.", "_", unmapped_names)
)

mapped_bam_lengths <- lapply(mapped_bams, function(x) {
  return(qwidth(x))
})


##################################################################################################################################
### 4. Combine and plot lengths ###
##################################################################################################################################

# chronologically order outputs:
all_lengths <- list(
  orig_fq = fq_lengths$orig_fq,
  unmapped_bam = unmapped_bam_lengths$unmapped_bam,
  unmapped_filtered_bam = unmapped_bam_lengths$unmapped_filtered_bam,
  filtered_orig_fq = fq_lengths$filtered_orig_fq,
  initial_mapped_bam = mapped_bam_lengths$initial_mapped_bam,
  initial_mapped_discordant_bam = mapped_bam_lengths$initial_mapped_discordant_bam,
  initial_mapped_split_bam = mapped_bam_lengths$initial_mapped_split_bam,
  consensus_unmapped_bam = unmapped_bam_lengths$consensus_unmapped_bam,
  consensus_fq = fq_lengths$consensus_fq,
  consensus_remapped_bam = mapped_bam_lengths$consensus_remapped_bam,
  consensus_remapped_discordant_bam = mapped_bam_lengths$consensus_remapped_discordant_bam,
  consensus_remapped_split_bam = mapped_bam_lengths$consensus_remapped_split_bam,
  filtered_consensus_bam = mapped_bam_lengths$filtered_consensus_bam,
  filtered_consensus_discordant_bam = mapped_bam_lengths$filtered_consensus_discordant_bam,
  filtered_consensus_split_bam = mapped_bam_lengths$filtered_consensus_split_bam
)

for (i in 1:length(all_lengths)) {

  png(paste0(stats_dir, i, ".", names(all_lengths)[i], "_read_length_distribution.png"))
    plot(density(all_lengths[[i]]))
  dev.off()

}








